### PR TITLE
Add an option to exclude development dependencies from the SBOM

### PR DESCRIPTION
**Pull request by andreas-hilti**
Add an option to exclude development dependencies from the SBOM

This aims to address https://github.com/CycloneDX/cyclonedx-conan/issues/85

This adds an  `--exclude-dev` command line option, which allows you to exclude development dependencies from the generated SBOM, for instance:
`cyclonedx-conan -l conan.lock conanfile.txt --exclude-dev`

### CODE REVIEW

File: src/command.py [hunk 1]
```diff
@@ -25,9 +25,10 @@
 from uuid import uuid4
 from conans.client.conan_api import Conan, ProfileData
 from conans.client.command import Command as ConanCommand, OnceArgument, Extender, _add_common_install_arguments
-from conans.client.graph.graph import DepsGraph
+from conans.client.graph.graph import DepsGraph, Node
```
[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py:28]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py:28]: (as this is a prexisting issue)

File: src/command.py [hunk 2]
```diff
@@ -121,9 +147,13 @@ def execute(self):
                 }
                 for dependency in node.dependencies:
                     purl = get_purl(dependency.dst.remote, dependency.dst.ref)
+                    if self._arguments.exclude_dev and str(dependency.dst.id) not in required_ids:
```
[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py]: Addressed it.

[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py]: Addressed it.

[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py]: Addressed it.

File: src/command.py [hunk 3]
```diff
@@ -110,6 +116,26 @@ def execute(self):
             'components': [],
             'dependencies': [],
         }
+
+        required_ids = set()
+        if self._arguments.exclude_dev:
+            to_visit: List[Node] = []
+            visited_ids = set()
+            node: Node
+            for node in deps_graph.nodes:
```
Comment (jkowalleck) [inline src/command.py:125]: can be shortened like 
```py
to_visit = set(node in deps_graph.nodes if node.ref is None)
```
now `to_visit` is a (unique) set, not a list, this saves comparisons later. 
you will be able to traverse it, simply.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: src/command.py (hunk 3 above)
Comment (andreas-hilti) [reply src/command.py]: I changed it as suggested.

File: src/command.py [hunk 4]
```diff
@@ -110,6 +116,22 @@ def execute(self):
             'components': [],
             'dependencies': [],
         }
+
+        required_ids = set()
+        if self._arguments.exclude_dev:
+            visited_ids = set()
+            node: Node
```
Comment (jkowalleck) [inline src/command.py:123]: this type hint should be moved where it is actually used - near 
```py
- node = to_visit.pop()
+ node: Node = to_visit.pop()
```

in addition with the following, a type hint for `node` would net be needed, right? 
```patch
+ from typing import Set
#...
- to_visit = set(node for node in deps_graph.nodes if node.ref is None)
+ to_visit: Set[Node] = set(node for node in deps_graph.nodes if node.ref is None)
```

File: src/command.py [hunk 5]
```diff
@@ -25,9 +25,10 @@
 from uuid import uuid4
 from conans.client.conan_api import Conan, ProfileData
 from conans.client.command import Command as ConanCommand, OnceArgument, Extender, _add_common_install_arguments
-from conans.client.graph.graph import DepsGraph
+from conans.client.graph.graph import DepsGraph, Node
 from conans.errors import ConanMigrationError, ConanException
 from packageurl import PackageURL
+from typing import List
```
[the remark this answers is not in the document]
Comment (andreas-hilti) [reply src/command.py]: You are correct, with the last change this is no longer used. Adapted it.

File: src/command.py (hunk 4 above)
Comment (andreas-hilti) [reply src/command.py]: I changed it. I hope sonatype-lift is now happy as well.

Comment (jkowalleck) [reply src/command.py]: thanks for applying the changes.


btw: 
lift was disabled, for reasons: https://community.sonatype.com/t/sunset-notice-sonatype-lift/10940

**Review by jkowalleck — APPROVED** [review]

**Pull request merged**